RavenDB-14974 introduce typed time-series session API

karmeli87 opened the pull request.

karmeli87 committed:
RavenDB-14974 allow to pass `null` for time-series policies

karmeli87 committed:
RavenDB-14974 introduce typed time-series session API

karmeli87 committed:
RavenDB-14974 add basic tests

karmeli87 committed:
RavenDB-14974 don't allow to mutate the values after append

karmeli87 committed:
RavenDB-14974 add time-series raw results to be typed

karmeli87 committed:
RavenDB-14974 add typed API for raw aggregation results

karmeli87 committed:
RavenDB-14974 add LINQ support

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/GetTimeSeriesOperation.cs:
```diff
@@ -104,9 +110,9 @@ public override void SetResponse(JsonOperationContext context, BlittableJsonRead
                     return;
 
-                Result = JsonDeserializationClient.TimeSeriesRangeResult(response);
+                Result = JsonDeserializationBase.GenerateJsonDeserializationRoutine<TimeSeriesRangeResult<TValues>>()(response);
```
Do we want to cache those routines? Now it generates them everytime?

karmeli87 replied:
Can we cache those? Will try but not sure it's possible because of `TValues`.

karmeli87 replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/TimeSeries/TimeSeriesRangeResult.cs:
```diff
@@ -10,4 +10,9 @@ public class TimeSeriesRangeResult
         internal string Hash;
     }
+
+    public class TimeSeriesRangeResult<TValues> : TimeSeriesRangeResult where TValues : TimeSeriesEntry
```
Would it help us to have it the other way around?

The `TimeSeriesRangeResult` would inherit from `TimeSeriesRangeResult<TimeSeriesEntry>`? Would that help us to avoid 'new' ?

karmeli87 replied:
We can't.
Because we can't cast from base class to the derived.

ppekrol commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentTimeSeries.cs:
```diff
@@ -546,5 +350,17 @@ private static IEnumerable<TimeSeriesEntry> ChopRelevantRange(TimeSeriesRangeRes
+
+        Task<IEnumerable<TValues>> IAsyncSessionDocumentTypedTimeSeries<TValues>.GetAsync(DateTime? @from, DateTime? to, int start, int pageSize, CancellationToken token)
+        {
+            throw new NotImplementedException();
```
Not implemented?

ppekrol commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentTimeSeries.cs:
```diff
@@ -546,5 +350,17 @@ private static IEnumerable<TimeSeriesEntry> ChopRelevantRange(TimeSeriesRangeRes
+
+        public void Append(TValues entry)
+        {
+            throw new NotImplementedException();
```
Not implemented?

ppekrol commented on src/Raven.Client/Documents/Session/AsyncSessionDocumentTimeSeries.cs:
```diff
@@ -546,5 +350,17 @@ private static IEnumerable<TimeSeriesEntry> ChopRelevantRange(TimeSeriesRangeRes
+            throw new NotImplementedException();
+        }
+
+        public void Append(TValues entry)
```
Implement that explicitly `IAsyncSessionDocumentTypedTimeSeries<TValues>.Append(TValues entry)`

karmeli87 replied:
deon

karmeli87 committed:
RavenDB-14974 implement missing async calls

ppekrol commented on src/Raven.Client/Documents/Session/SessionDocumentTimeSeries.cs:
```diff
@@ -35,11 +35,21 @@ public void Append(DateTime timestamp, double value, string tag = null)
 
+        IEnumerable<TValues> ISessionDocumentTypedTimeSeries<TValues>.Get(DateTime? from, DateTime? to, int start, int pageSize)
+        {
+            return AsyncHelpers.RunSync(() => _asyncSessionTimeSeries.GetAsyncInternal<TValues>(from, to, start, pageSize));
```
Shouldn't this call `GetAsync` not `GetAsyncInternal` ?

karmeli87 replied:
`GetAsync` is only for `TimeSeriesEntry`, here you have `TValues`

karmeli87 committed:
RavenDB-14974 cache routines for range parsing

karmeli87 committed:
RavenDB-14974 Address PR comments

karmeli87 committed:
RavenDB-14974 Fix `IsTimeSeriesCall` recognition

karmeli87 committed:
RavenDB-14974 Fix studio compilation

karmeli87 committed:
RavenDB-14974 Create seperate class for aggregated result and raw result

karmeli87 committed:
RavenDB-14974 remove initial value from Raven_Max and Raven_Min

An empty array is already propogated as null

This change is needed to have time series aggregated as double[] instead of double?[].

ppekrol commented on src/Raven.Client/Documents/Session/TimeSeries/TimeSeriesEntry.cs:
```diff
@@ -5,17 +5,31 @@
-        public double Value => Values[0];
+    }
+
+    public abstract class TimeSeriesValues
```
TimeSeriesEntryValues?

karmeli87 committed:
RavenDB-14974 address PR comments
